pkg/virt-controller/watch/vm.go

```diff
@@ -271,7 +276,94 @@ func (c *VMController) execute(key string) error {
 		return err
 	}
 
-	return createErr
+	if createErr != nil {
+		return createErr
+	}
+
+	return nil
+}
+
+// Handles VM rename create requests
+// First return value is a boolean indicating if the controller should retry the request
+func (c *VMController) handleVMRenameCreateRequest(vm *virtv1.VirtualMachine, newName string) (bool, error) {
+	// Check if a VM named 'newName' exists
+	_, err := c.clientset.VirtualMachine(vm.Namespace).Get(newName, &v1.GetOptions{})
+
+	if err != nil {
+		// If the VM existence could not be determined, retry
+		if !strings.Contains(err.Error(), "not found") {
+			return true, err
+		} else {
+			// If the VM was not found, we can create it
+			newVM := vm.DeepCopy()
+
+			newVM.ResourceVersion = ""
+			newVM.Name = newName
+
+			// Update the VM label if it exists
+			if newVM.Labels != nil {
+				_, hasVMLabel := newVM.Labels[virtv1.VirtualMachineLabel]
+
+				if hasVMLabel {
+					newVM.Labels[virtv1.VirtualMachineLabel] = newName
+				}
+			}
+
+			// Update the VMI spec VM label if it exists
+			if newVM.Spec.Template.ObjectMeta.Labels != nil {
+				_, hasVMLabel := newVM.Spec.Template.ObjectMeta.Labels[virtv1.VirtualMachineLabel]
+
+				if hasVMLabel {
+					newVM.Spec.Template.ObjectMeta.Labels[virtv1.VirtualMachineLabel] = newName
+				}
+			}
+
+			// Clear VM status
+			newVM.Status = virtv1.VirtualMachineStatus{}
+
+			// Add a condition to the new VM to tell the controller it was renamed
+			newVM.Status.Conditions = []virtv1.VirtualMachineCondition{
+				{
+					Type:    virtv1.RenameConditionType,
+					Status:  "Successful",
+					Reason:  vm.Name,
+					Message: fmt.Sprintf("This VM was renamed, the old name was %s", vm.Name),
+				},
+			}
+
+			// Add a RenameDeleteRequest to the new VM to remove the VM with the old Name
+			newVM.Status.StateChangeRequests = []virtv1.VirtualMachineStateChangeRequest{
+				{
+					Action: virtv1.RenameDeleteRequest,
+					Data: map[string]string{
+						"oldName": vm.Name,
+					},
+				},
+			}
+
+			// Attempt creation of the new VM
+			_, err = c.clientset.VirtualMachine(vm.Namespace).Create(newVM)
+
+			if err != nil {
+				return true, err
+			}
+		}
+	}
+
+	return false, fmt.Errorf("a VM with the new name was found, rejecting rename request")
+}
+
+// Handles VM rename delete requests
+// First return value is a boolean indicating if the controller should retry the request
+func (c *VMController) handleVMRenameDeleteRequest(vm *virtv1.VirtualMachine, oldName string) (bool, error) {
```

Comment:
I think we need additional security here.  Imagine a user with RBAC permission limited to a single VM in a namespace. 
 Correct me if I'm wrong, but is not possible for that user to add a `RenameDeleteRequest` to the status of that vm allowing them to delete any arbitrary VM in the namespace?

Reply:
The specific scenario you described is not possible because of
https://github.com/kubevirt/kubevirt/blob/8a51882bf5754c497579bddee9be236c693a9ba4/pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go#L253
Which prevents updating a VM with a `RenameDeleteRequest`.
The other issue that I found is that a user can **create** a new VM with a `RenameDeleteRequest` and delete an arbitrary VM in the namespace, which is still bad because it means that `create == delete` permission wise.
In order to make the entire process safe and failure proof, I might need to utilize a config map that will hold the current status of the renaming process and take that into consideration every time the controller notices the VM that has to be renamed.

Reply:
I would like to know your opinion on this solution which I think would solve the issue:
https://github.com/kubevirt/kubevirt/pull/3036#discussion_r381870477

Reply:
I think it may make sense verify a field in the vm before it is deleted.  However, don't you possibly have a chicken/egg situation or race condition?  Isn't the `RenameCreateRequest` removed from the source vm right after the new vm is created?

Reply:
You are right, I think that was the original reason why I haven't checked for a field in the old vm before (I knew I had a reason for not doing that, but forgot what it was).
I'll think about it some more and try coming up with a proper solution